### PR TITLE
Email 2fa

### PR DESCRIPTION
**Pull request by pickfire**
Revive #267 thanks to @Atterratio for the initial patch. I did some changes based on the reviews on the original pull requests and some doc style fixes.

## Description

Make possible two-factor authorization using emails.

## Motivation and Context

Although this way of two-factor authentication is not very reliable, it can be preferable for some users.

## How Has This Been Tested?

Tested in a local project. Not fully tested but I tried it out. Didn't look closely at old PR.

## Screenshots (if appropriate):
```
Content-Type: multipart/alternative;
 boundary="===============1209054917869947523=="
MIME-Version: 1.0
Subject: Authentication token email
From: hello@test.org
To: hello@example.com
Date: Wed, 28 Jul 2021 04:20:20 -0000
Message-ID: <162744602025.73619.15432270616108599468@mi.localdomain>

--===============1209054917869947523==
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit

Hello,
Your email address has been given for two-factor authorization on the our website.
If you did't do this, just ignore this message.

Authentication token for user hello is 487008.
--===============1209054917869947523==
Content-Type: text/html; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit



<p>Hello,</p>
<p>
    
        Your email address has been given for two-factor authorization on the our website.
        If you did't do this, just ignore this message.
    
</p>
<p>Authentication token for user hello is 487008.</p>
--===============1209054917869947523==--

-------------------------------------------------------------------------------
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by WayneLambert** [comment]
Hi @pickfire,

Thanks for reviving this and continuing the effort from #267. I am in the midst of considering bolting on a custom implementation using email, however would prefer to adjust/extend the use of this package, so if this is merged, it would greatly help.

It looks like a review is required and is the blocker at the moment. Perhaps you could add @Atterratio as the reviewer?

Thanks again!

**Comment by pickfire** [comment]
> It looks like a review is required and is the blocker at the moment. Perhaps you could add @Atterratio as the reviewer?

I could but I don't think he will review it since he is not active in the original patch as well.

**Comment by WayneLambert** [comment]
> I cannot since I don't have permission but I don't think he will review it since he is not active in the original patch as well.

Hi @pickfire, thanks for this.

Perhaps @moggers87 could take a look at this since it is a continuation from the previous PRs work (#267) and seems the most likely contributor to know what is required to get this over the line.

**Comment by moggers87** [comment]
I have requested a review from @moggers87

**Comment by daronzwink** [comment]
This is awesome, and exactly what we were looking for! Thank you very much for picking this enhancement back up and running with it. We would love to apply this to our production environment and wanted to see if there is any chance this could be reviewed and merged?

**Comment by pickfire** [comment]
I am waiting for reviewer, not sure if the maintainers are still active.

File: two_factor/models.py
Comment (moggers87) [inline two_factor/models.py:71]: `drift_range` should be set to `None` here. That way if someone forgets to set it, they'll get an error.

File: two_factor/models.py
Comment (moggers87) [inline two_factor/models.py:96]: This should not be in the mixin and there should be a `TWO_FACTOR_EMAIL_THROTTLE_FACTOR` setting for `EmailDevice`

**Review by moggers87 — CHANGES_REQUESTED** [review]
A few changes required, otherwise good :+1:

**Comment by claudep** [comment]
I hoped that #331 would be merged before any new capability addition. The more we wait the more it will be difficult to change the module structure.

**Comment by moggers87** [comment]
@claudep that's a slightly bigger change, which is why I keep delaying merging it :laughing:

**Comment by moggers87** [comment]
Actually nothing can get merged yet anyway because Travis is down and I need to complete #425 to fix that.

**Comment by J4bbi** [comment]
I am also interested in email support. And I am guessing it's a feature that would be in demand.

Can I ask what is holding this PR up at the moment? Original PR is from 2018.

@moggers87  it looks like @pickfire has applied your requested changes (admit I haven't looked at the code)

Do you have time for this?

As things stand we might try to implement the PR directly ourselves but we would be more relaxed about it if this PR was merged :) 

Thanks to everyone for their work on this.

**Comment by giovannicimolin** [comment]
@moggers87 (CC @pickfire) I'm also interested in having this merged. 
Is there anything I can do to help speed up this process? 
Are there any blockers on this?

**Comment by claudep** [comment]
FYI I'm currently trying to refresh this patch based on the new plugin-based method.

**Comment by claudep** [comment]
The new version is now in #475.

**Pull request closed without merging**
